### PR TITLE
[RFC] Clean up docs and clint warnings ex_cmds2.c

### PR DESCRIPTION
**Pull request by phanimahesh**
Had these lying around since long, left over from #4289.
Might as well see it to completion by removing all warnings
and updating all docs.

### CODE REVIEW

**Comment by phanimahesh** [comment]
Before I put in more efforts on this, will this be welcome? It will be a gainormous diff, with purely documentation updates and stylistic changes to silence clint. However having updated documentation will be a huge plus for anyone who is getting started with the codebase.

There are two options to review this:
1. Use tools on the local machine and handle this as a single large diff
2. cherry-pick commits from here to master when a hunk is satisfactory, aggressively rebasing this ongoing PR

Option 2 keeps things simple. However I understand if it isn't an option or if there is no bandwidth for this right now.

**Comment by justinmk** [comment]
> Might as well see it to completion by removing all warnings and updating all docs.

Given the aggressiveness of clint, I think a PR like this is probably OK. But please put the "removing all warnings" changes in a separate commit from the style changes.

**Comment by phanimahesh** [comment]
Thanks! I'll rewrite the commits to group clint changes and doc and other updates separately.

Github interface does not show large diffs by default. I'm guessing handling clint will be very straightforward from my experience, but I would like to receive comments on the doc updates.

Would it be possible to handle this incrementally as a set of PRs each dealing with few hundred lines and the cascading changes? The idea is to keep diffs reasonably sized so that github doesn't suppress them and reviews can be expected on github directly.

**Comment by justinmk** [comment]
> Would it be possible to handle this incrementally as a set of PRs each dealing with few hundred lines and the cascading changes?

Separate commits in the same PR can serve the same purpose.

**Comment by phanimahesh** [comment]
I need help with https://github.com/neovim/neovim/blob/7a1f6516d106d56103ae9fa1aec56e7a689ad3b2/src/nvim/ex_cmds2.c#L3241-L3274

How can I add braces that don't change `#ifdef` semantics for all use cases?

**Comment by ZyX-I** [comment]
@phanimahesh I thin you can do the following:

```Patch
--- braces-orig.c	2016-04-11 19:46:43.963281960 +0300
+++ braces.c	2016-04-11 19:46:11.595682038 +0300
@@ -4,9 +4,12 @@
 #ifdef HAVE_WORKING_LIBINTL
     if (what == VIM_LC_MESSAGES) {
       p = get_mess_env();
-    } else
+    } else {
 #endif
       p = (char_u *)setlocale(what, NULL);
+#ifdef HAVE_WORKING_LIBINTL
+    }
+#endif
     if (p == NULL || *p == NUL) {
       p = (char_u *)"Unknown";
     }
@@ -15,15 +18,16 @@
 #ifndef LC_MESSAGES
     if (what == VIM_LC_MESSAGES) {
       loc = "";
-    } else
+    } else {
 #endif
-    {
       loc = setlocale(what, (char *)name);
 #ifdef LC_NUMERIC
       // Make sure strtod() uses a decimal point, not a comma.
       setlocale(LC_NUMERIC, "C");
 #endif
+#ifdef LC_MESSAGES
     }
+#endif
     if (loc == NULL) {
       EMSG2(_("E197: Cannot set language to \"%s\""), name);
     } else {
```

(Note: I have constructed a small file for test, so offsets in diff are definitely wrong by a huge amount of lines.) It looks like I have introduced two more `#ifdef`s which is not very good, but code should act the same and not fail lint check.

**Comment by phanimahesh** [comment]
Ah, I haven't thought of introducing additional ifdefs. Thanks @ZyX-I !.

Aside: I really don't like those ifdefs in there, but doesn't look like there is any non-invasive way of removing them. I may revisit these conditionally compiled blocks later.

**Comment by phanimahesh** [comment]
clint recommends changing `strtok` to `os_strtok_r`. I instead changed it to `strtok_r` which is also threadsafe. Currently added a NOLINT to suppress this warning. Is that OK?

If not, how do I go about implementing `os_strtok_r`?

Also, I'm done with code-wise changes. I automated a lot of it using vim macros, `astyle` also helped. Now to the next part - documentation updates.

Github is already refusing to show diffs, and if I continue the kind of splitting I started we will easily end up with hundreds of tiny commits. Any suggestions on splitting the commits for soliciting reviews?

**Comment by bfredl** [comment]
ain't strtok_r already threadsafe ? Then you could change `clint.py` to map `strtok` to `strtok_r` instead and remove `os_strtok_r`

**Comment by ZyX-I** [comment]
@phanimahesh @bfredl `strtok_r` is POSIX. `strtok` is C99. We assume C99 support, but not POSIX support, so `os_strtok_r` and not `strtok_r`.

**Comment by phanimahesh** [comment]
@ZyX-I In that case, how do I implement os_strtok_r? My limited search turned up nothing useful.

Also, TIL vim/neovim doesn't assume POSIX support. Out of curiosity, what platforms do we support that aren't POSIX compliant?

**Comment by bfredl** [comment]
windows, seems its called strtok_s on windows: https://msdn.microsoft.com/en-us/library/ftsafwz3.aspx

**Comment by phanimahesh** [comment]
So something along the lines of
```c
char* os_strtok_r(char* str, const char* delim, char** savestate)
{
#if LL_WINDOWS
	return strtok_s(str, delim, savestate);
#else
	return strtok_r(str, delim, savestate);
#endif
}
```
should work, right? I lifted it from http://kirstensviewer.googlecode.com/svn/trunk/indra/llcorehttp/_httpoprequest.cpp, the only useful search result I found.

**Comment by bfredl** [comment]
Yes, though the condition should be WIN32. Though I wonder if it would hurt to just `#define` it.

**Comment by phanimahesh** [comment]
btw, why did the appveyor build succeed when I used `strok_r` everywhere? It also tries a win32 build, right?

I'll go with `#define`ing them since that is easier on the eyes. Dropping them at the top of file after `#include`s for now. Will move later if required.

File: src/nvim/ex_cmds2.c [hunk 1]
```diff
@@ -868,7 +868,7 @@ void set_context_in_profile_cmd(expand_T *xp, char_u *arg)
     return;
   }
 
-  // TODO: expand function names after "func
+  // TODO: expand function names after "func // NOLINT(readability/todo)
```
Comment (ZyX-I) [inline src/nvim/ex_cmds2.c]: Just bisect to determine who added this TODO.

File: src/nvim/ex_cmds2.c [hunk 2]
```diff
@@ -3324,7 +3324,7 @@ static char_u **find_locales(void)
 {
   garray_T locales_ga;
   char_u      *loc;
-  char *strtok_r_save_ptr = NULL;
+  char *saveptr = NULL;
 
   // Find all available locales by running command "locale -a".  If this
   // doesn't work we won't have completion. */
```
Comment (ZyX-I) [inline src/nvim/ex_cmds2.c]: I did not see the whole diff, but I assume it also contains `*/`. These should not be here.

File: src/nvim/ex_cmds2.c (hunk 1 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: You must have meant blame. I blamed it repeatedly and found that it was introduced at 72cf89bce8e4230dbc161dc5606f48ef9884ba70 by @tarruda.

I'm marking it `TODO(Thiago de Arruda): ...` Is that okay?

File: src/nvim/ex_cmds2.c (hunk 2 above)
Comment (phanimahesh) [inline src/nvim/ex_cmds2.c]: Thanks for pointing it out, I must have missed some. I'll go through the whole diff again and remove these and any thing else I have missed.

File: src/nvim/ex_cmds2.c (hunk 1 above)
Comment (ZyX-I) [inline src/nvim/ex_cmds2.c]: @phanimahesh Yes, annotate. Use nick: I see `TODO(tarruda)` in some files (e.g. os/input.c:88).

**Comment by phanimahesh** [comment]
builds failing due to multiple reasons.

1. New lint checks added recently. I'll address them.
2. Some issue with lua scripts.
```
busted: error: Cannot load helper script: /home/travis/build/neovim/neovim/test/functional/preload.lua

./test/functional/helpers.lua:5: module 'nvim.child_process_stream' not found:No LuaRocks module found for nvim.child_process_stream
```
3. Something saying unable to find TAP stream on @jszakmeister's quickbuilds. I don't understand this one.

2 is reproducible locally on master.

**Comment by phanimahesh** [comment]
Docs can be improved still, but with recent vim patches, maintaining this mergeable with master is taking effort. Bumping to RFC.

 - The only code changes are int->bool where appropriate, pre->post {in,de}crement throughout, and adding braces around all blocks where needed.
 - All comments moved to `//` style from `/* ... */`
 - Where function args have been annotated with comments, moved them to `@param` s

Other than these, the changes are minimal and only in comments.

**Comment by phanimahesh** [comment]
Also, sorry @justinmk, separating into multiple commits while keeping clint happy with all of them was becoming a little inconvenient, and so I squashed.

**Comment by phanimahesh** [comment]
Ping. Comments? Doing yet another rebase, though thankfully this one has no conflicts.

**Comment by KillTheMule** [comment]
There's one new linter error, maybe you could fix that up?

`src/nvim/ex_cmds2.c:2343:  Should leave no spaces after a cast: u'(void *) &cookie'  [whitespace/cast] [2]`

**Comment by phanimahesh** [comment]
New clint rules added forbidding spaces after cast. Corrected for them.

**Comment by justinmk** [comment]
Continued at https://github.com/neovim/neovim/pull/4965. See comments there.

**Pull request closed without merging**